dsc/base.py

```diff
@@ -0,0 +1,96 @@
+from __future__ import annotations
+
+import logging
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, Any, final
+
+from dsc.item_submission import ItemSubmission
+from dsc.utilities.aws.s3 import S3Client
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+logger = logging.getLogger(__name__)
+
+
+class BaseWorkflow(ABC):
+    """A base workflow class from which other workflow classes are derived."""
+
+    def __init__(
+        self,
+        email_recipients: list[str],
+        metadata_mapping: dict,
+        s3_bucket: str,
+        s3_prefix: str | None,
+    ) -> None:
+        """Initialize base instance.
+
+        Args:
+            email_recipients: The email addresses to notify after runs of the workflow.
+            metadata_mapping: A mapping file for generating DSpace metadata from the
+            workflow's source metadata.
+            s3_bucket: The S3 bucket containing bitstream and metadata files for the
+            workflow.
+            s3_prefix: The S3 prefix used for objects in this workflow. This prefix does
+            NOT include the bucket name.
+        """
+        self.email_recipients: list[str] = email_recipients
+        self.metadata_mapping: dict = metadata_mapping
+        self.s3_bucket: str = s3_bucket
+        self.s3_prefix: str | None = s3_prefix
+
+    @final
+    def generate_submission_batch(self) -> Iterator[tuple[str, list[str]]]:
+        """Generate a batch of item submissions for the DSpace Submission Service.
+
+        MUST NOT be overridden by workflow subclasses.
+        """
+        s3_client = S3Client()
+        batch_metadata = self.get_batch_metadata()
+        for item_metadata in batch_metadata:
+            item_identifier = self.get_item_identifier(item_metadata)
```

Comment:
I would say every method the the workflow classes are tightly coupled and opinionated to that workflow. In retrospect, I thought I errored in typing `item_metadata=Any` in `get_item_identifier` and `source_metadata=dict[str, Any]` in `ItemSubmission` since I didn't expect that object to be transformed between getting the identifier and instantiating `ItemSubmission`. 

However! That is now suggesting to me that a `get_item_metadata(batch_metadata)` method is needed and that those type hints should remain:

```
batch_metadata = self.get_batch_metadata()
for item_metadata in self.get_item_metadata(batch_metadata):
      item_identifier = self.get_item_identifier(item_metadata)
```

Currently, `batch_metadata` in most workflows is a CSV so `get_item_metadata` would be a simple `dict` iterator.  For `OpenCourseWare`, it's a list of zip files, so `get_item_metadata` iterates and extracts a `dict` from each zip file. If a future source uses `XML`, `get_item_metadata` would iterate and generate a `dict` from the `XML`. Thoughts?

Reply:
So would it be accurate to say?

1. `get_batch_metadata()` --> retrieves metadata from somewhere, but what is returned still contains multiple items; probably most often downloading from S3, but theoretically could get it other ways too (e.g. an HTTP request to ETD... 😎 ).  The result might be a single CSV reader/dataframe, a single parsed XML file object, etc.

2. `get_item_metadata()` --> would yield actual dictionaries of item metadata, for each item, doing whatever is needed (e.g. unzipping files, parsing XML, looping through CSV rows, etc.)

I'm having trouble articulating what feels confusing about the relationship of,
- `get_batch_metadata`
- `get_item_identifier`
- `get_bitstream_uris`
- and now the proposed `get_item_metadata`

It feels kind of like they are very tightly coupled, and creating new workflows that extend this class would be jumping through hoops to define those inter-related methods.  When -- and maybe missing something here -- a workflow could just be responsible for defining something like `fetch_items()` which would yield nearly everything you need to instantiate an `ItemSubmission`.  

Which, and I'm feeling dizzy from circular questioning, why wouldn't they _directly_ yield an `ItemSubmission` then?  

Each of those workflows could, and _should_, define helper methods as needed to support this work.  But in some cases, like reading a CSV from S3, it feels like `get_batch_metadata` (reading the CSV) and yielding items `get_item_metadata` could be achieved with a couple lines like:

```python
import pandas as pd

df = pd.read_csv("s3://....")
for _, row in df.iterrows():
   yield row # <----- basically an Item
```

Sorry if confusing the issue, and happy to hold off on these questions until more is established.